python/cudf/cudf/_lib/copying.pyx

```diff
@@ -130,7 +131,16 @@ def copy_range(Column input_column,
 
 
 def gather(Table source_table, Column gather_map, bool keep_index=True):
-    assert pd.api.types.is_integer_dtype(gather_map.dtype)
+    if not pd.api.types.is_integer_dtype(gather_map.dtype):
+        raise ValueError("Gather map is not integer dtype.")
+
+    if len(gather_map) > 0:
+        gm_min, gm_max = minmax(gather_map)
+        if gm_min < -len(source_table) or gm_max >= len(source_table):
+            raise IndexError(f"Gather map index with min {gm_min},"
+                             f" max {gm_max} is out of bounds in"
+                             f" {type(source_table)} with {len(source_table)}"
+                             f" rows.")
```

Comment:
How was this previously being checked? This makes every gather synchronize.

Reply:
In libcudf. See discussion here: https://github.com/rapidsai/cudf/issues/5505